azkaban-executor/azkaban-emr-jobtype/src/main/java/uk/gov/dwp/dataworks/azkaban/jobtype/EMRStep.java

```diff
@@ -320,38 +327,37 @@ protected String getClusterId(AmazonElasticMapReduce emr) {
                 try {
                     payload = new ObjectMapper().writeValueAsString(batchConfig);
                 } catch (Exception e) {
-                    error(e.getMessage());
+                    log.error(e.getMessage());
                     throw new IllegalStateException(e);
                 }
 
-                AWSLambda client = AWSLambdaClientBuilder.defaultClient();
                 InvokeRequest req = new InvokeRequest().withFunctionName("aws_analytical_env_emr_launcher")
                                                        .withPayload(payload);
 
                 InvokeResult result = client.invoke(req);
                 invokedLambda = true;
 
                 if (result.getStatusCode() != 200) {
-                    error(result.getFunctionError());
+                    log.error(result.getFunctionError());
                     throw new IllegalStateException(result.getFunctionError());
                 }
             }
 
-            if (clusterId == null) {
+            if (!clusterId.isPresent()) {
                 maxAttempts--;
                 try {
                     Thread.sleep(pollTime);
                 } catch (Exception e) {
-                    warn("Sleep interrupted");
+                    log.warn("Sleep interrupted");
                 }
             }
         }
 
-        if (!killed && clusterId == null) {
+        if (!killed.get() && !clusterId.isPresent()) {
             throw new IllegalStateException("No batch EMR cluster available");
         }
 
-        return clusterId;
+        return clusterId.orElse("");
```

Comment:
So we could be returning "" if the job is killed (which is caught after we call this method), or cluster hasn't started after our polling time. Shouldn't we be throwing an error rather than returning "" as the cluster name in that case?

Reply:
If the cluster hasn't appeared after the polling has ended then the IllegalStateException on line 357 is thrown. There is a test for this (line 176 EMRStepTest). If the job is killed then an empty string is returned and this is dealt with on line 150 of EMRStep.

Reply:
You're right! I'm just not a fan of the orElse part as it gives the impression its a valid value to return!